query-algebrizer/src/validate.rs

```diff
@@ -229,4 +274,177 @@ mod tests {
             _ => panic!(),
         };
     }
+
+
+    /// Tests that the top-level form is a valid `not`, returning the clauses.
+    fn valid_not_join(parsed: FindQuery, expected_unify: UnifyVars) -> Vec<WhereNotClause> {
+        // filter all the not clauses
+        let mut nots = parsed.where_clauses.iter().filter(|&x| match *x {
+            WhereClause::NotJoin(_) => true,
+            _ => false,
+        });
+
+        // there should be only one not clause
+        let clause = nots.next().unwrap().clone();
+        assert_eq!(None, nots.next());
+
+        match clause {
+            WhereClause::NotJoin(not_join) => {
+                // It's valid: the variables are the same in each branch.
+                assert_eq!((), validate_not_join(&not_join).unwrap());
+                assert_eq!(expected_unify, not_join.unify_vars);
+                not_join.clauses
+            },
+            _ => panic!(),
+        }
+    }
+
+    /// Test that a `not` is valid if all of its arms refer to the same variables.
+    #[test]
+    fn test_success_not() {
+        let query = r#"[:find ?name
+                        :where [?id :artist/name ?name]
+                            (not [?id :artist/country :country/CA]
+                                 [?id :artist/country :country/GB])]"#;
+        let parsed = parse_find_string(query).expect("expected successful parse");
+        let clauses = valid_not_join(parsed, UnifyVars::Implicit);
+
+        // Check each part of the body
+        let mut parts = clauses.into_iter();
+        match (parts.next(), parts.next(), parts.next()) {
+            (Some(clause1), Some(clause2), None) => {
+                assert_eq!(
+                    clause1,
+                    WhereNotClause::Clause(
+                        WhereClause::Pattern(Pattern {
+                            source: None,
+                            entity: PatternNonValuePlace::Variable(Variable::from_valid_name("?id")),
+                            attribute: ident("artist", "country"),
+                            value: value_ident("country", "CA"),
+                            tx: PatternNonValuePlace::Placeholder,
+                        }),
+                    ));
+                assert_eq!(
+                    clause2,
+                    WhereNotClause::Clause(
+                        WhereClause::Pattern(Pattern {
+                            source: None,
+                            entity: PatternNonValuePlace::Variable(Variable::from_valid_name("?id")),
+                            attribute: ident("artist", "country"),
+                            value: value_ident("country", "GB"),
+                            tx: PatternNonValuePlace::Placeholder,
+                        }),
+                    ));
+            },
+            _ => panic!(),
+        };
+    }
+
+
+    /// Test that a `not` with differing variable sets in each body part will fail to validate.
```

Comment:
This is invalid, but not for this reason. It's invalid because there's no binding for `?release` by the time the `not` is executed.

See this DataScript test for an example:

https://github.com/tonsky/datascript/blob/master/test/datascript/test/query_not.cljc#L183

(Strictly speaking there is an interpretation for this query: it's "find artist names for non-Canadian artists, if no albums were released by anybody in 1970". We prefer you to be explicit about that interpretation by using `not-join`!)